File: sdk/go/streaming.go
package llmengine

import (
	"bufio"
	"encoding/json"
	"io"
	"strings"
)

type ChatCompletionChunk struct {
	ID      string `json:"id"`
	Object  string `json:"object"`
	Created int64  `json:"created"`
	Model   string `json:"model"`
	Choices []struct {
		Index int `json:"index"`
		Delta struct {
			Content string `json:"content"`
		} `json:"delta"`
		FinishReason *string `json:"finish_reason"`
	} `json:"choices"`
}

type StreamResponse struct {
	Chunk ChatCompletionChunk
	Err   error
}

type Stream struct {
	reader *bufio.Reader
	body   io.ReadCloser
}

func NewStream(body io.ReadCloser) *Stream {
	return &Stream{
		reader: bufio.NewReader(body),
		body:   body,
	}
}

func (s *Stream) Recv() (ChatCompletionChunk, error) {
	for {
		line, err := s.reader.ReadString('\n')
		if err != nil {
			s.body.Close()
			return ChatCompletionChunk{}, err
		}

		line = strings.TrimSpace(line)
		if line == "" || strings.HasPrefix(line, ":") {
			continue
		}

		if strings.HasPrefix(line, "data: ") {
			data := strings.TrimPrefix(line, "data: ")
			if data == "[DONE]" {
				s.body.Close()
				return ChatCompletionChunk{}, io.EOF
			}

			var chunk ChatCompletionChunk
			if err := json.Unmarshal([]byte(data), &chunk); err != nil {
				continue
			}
			return chunk, nil
		}
	}
}

func (s *Stream) Close() error {
	return s.body.Close()
}

File: sdk/typescript/src/streaming.ts
export interface ChatCompletionChunk {
  id: string;
  object: string;
  created: number;
  model: string;
  choices: Array<{
    index: number;
    delta: { content?: string };
    finish_reason: string | null;
  }>;
}

export async function* streamIterator(response: Response): AsyncGenerator<ChatCompletionChunk, void, unknown> {
  if (!response.body) {
    throw new Error("Response body is null or undefined.");
  }

  const reader = response.body.getReader();
  const decoder = new TextDecoder("utf-8");
  let buffer = "";

  try {
    while (true) {
      const { done, value } = await reader.read();
      if (done) break;

      buffer += decoder.decode(value, { stream: true });
      const lines = buffer.split("\n");
      buffer = lines.pop() || "";

      for (const line of lines) {
        const trimmed = line.trim();
        if (!trimmed || trimmed.startsWith(":")) continue;

        if (trimmed.startsWith("data: ")) {
          const dataStr = trimmed.slice(6).trim();
          if (dataStr === "[DONE]") return;

          try {
            const chunk: ChatCompletionChunk = JSON.parse(dataStr);
            yield chunk;
          } catch (err) {
            // Ignore parse errors on transient chunks
          }
        }
      }
    }
  } finally {
    reader.releaseLock();
  }
}
